src/transaction/log_prior_recv.cpp

```diff
@@ -58,11 +59,50 @@ namespace cublog
     std::unique_lock<std::mutex> ulock (m_messages_mutex);
     m_shutdown = true;
     ulock.unlock ();
-    m_messages_condvar.notify_one ();
+    m_messages_push_cv.notify_one ();
 
     m_thread.join ();
   }
 
+  void
+  prior_recver::wait_until_empty_and_pause ()
+  {
+    std::unique_lock<std::mutex> ulock (m_messages_mutex);
+
+    assert (m_paused == false);
+    assert (m_shutdown == false);
+
+    constexpr auto sleep_30ms = std::chrono::milliseconds (30);
+    while (!m_messages_consume_cv.wait_for (ulock, sleep_30ms, [this]
+    {
+      return m_messages.empty();
+      }));
+
+    m_paused = true;
+  }
+
+  void
+  prior_recver::resume ()
+  {
+    std::unique_lock<std::mutex> ulock (m_messages_mutex);
+
+    assert (m_paused == true);
+    assert (m_shutdown == false);
+
+    m_paused = false;
+
+    if (!m_messages.empty ())
```

Comment:
Nit.
`notify_one` from within the lock?
I would enclose the messages mutex in a scope; in the locked scope save a bool flag whether to notify or not; notify if needed outside the scope:
```
  bool notify_one = false;
  { // locked scope
    std::unique_lock<std::mutex> ulock (m_messages_mutex);
    // ...
    notify_one = !m_messages.empty ();
    // ...
  }
  if (notify_one)
    {
      // notify_one
    }
```

Same remark/question for `m_messages_consume_cv.notify_one ();` below.